### PR TITLE
Support for BaseAudioContext.currentTime and state changes

### LINKED ISSUE #22: Implement support for the BaseAudioContext.currentTime property
[BaseAudioContext.currentTime](https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-currenttime)

### PR DESCRIPTION
**Pull request by ferjm**
Fixes #22 and #23

### CODE REVIEW

File: src/audio/graph.rs [hunk 1]
```diff
@@ -2,26 +2,46 @@ use audio::node::{AudioNodeMessage, AudioNodeType};
...
 use std::sync::mpsc::{self, Sender};
+use std::sync::{Arc, RwLock};
 use std::thread::Builder;
 
 static NEXT_NODE_ID: AtomicUsize = ATOMIC_USIZE_INIT;
 
 pub struct AudioGraph {
     sender: Sender<AudioRenderThreadMsg>,
+    sample_rate: f32,
+    current_time: Arc<RwLock<f64>>,
```
Comment (Manishearth) [inline src/audio/graph.rs]: does the AudioGraph need this? I'd rather have this be owned by the actual render thread without an RWLock

Comment (Manishearth) [reply src/audio/graph.rs]: overall we should clean up the mutability story here but I have a plan for that

File: src/audio/render_thread.rs [hunk 2]
```diff
@@ -117,7 +121,12 @@ impl AudioRenderThread {
                 }
                 // and push into the audio sink the result of processing a
                 // render quantum.
-                let _ = self.sink.push_data(self.process());
+                if let Ok(duration) = self.sink.push_data(self.process()) {
```
Comment (Manishearth) [inline src/audio/render_thread.rs]: we shouldn't need to ask the sink for the duration; we know how long it is. The sink calculates the duration from information on the buffer that _we_ provided.

File: src/audio/graph.rs (hunk 1 above)
Comment (ferjm) [reply src/audio/graph.rs]: The WebAudio API exposes sync getters for these values, so I figured that we should make the control thread own them. I guess we can have a sync channels with buffer size 0 for these getters (?).

**Comment by Manishearth** [comment]
Perhaps yeah.

**Comment by Manishearth** [comment]
I don't think we need a sync sender here, we can use the same one -- just block on the receiver end.

**Pull request merged**